### ISSUE TITLE
Full repository cannot be used in Windows because standard Windows filenames are not case sensitive

### ISSUE THREAD

**Issue by fho4abcd** [ABCD-DEVCOM/ABCD2#7]
Unpacking a zip warns about duplicates and pathnames that er too long

GitHub Desktop on windows accepts this but may conclude that the file is changed.

**Preliminary solution: use GitHub Desktop only on Linux.**

**Unzip at the top of the filesystem to avoid pathnames that are too long**

Best solution would be Windows with case-sensitive names and longer pathnames. A short survey did not reveal a good workaround.
Inspection of some found duplicates show that it is also possible to delete one of them. Easy examples:
- bases-examples_Linux\agrovoc\pfts\en\rt.pft -> RT.pft differ in line endings only
- bases-examples_Linux\marc\def\pt\countries.tab -> .TAB has more entries=not used. .tab of windows is equal to linux
- bases-examples_Linux\marcuni\def\pt\countries.tab -> .TAB are equal=not used
- bases-examples_Linux\reserve\marc0\def\pt\countries.TAB -> .TAB has more entries=not used. no windows equivalent
- bases-examples_Linux\stock\pfts\en\checkbarcodesMFN.pft -> very different mfn is not used
- bases_examples_Windows\stock\pfts\en\checkbarcodesMFN.pft-> very different mfn is not used
More problematic:
- dhtmlxgridCell.js. -> Many variants exist (find . -name dhtml*rid*ell.js -exec ls -l {} \; ). Grep delivers usage information (grep -iR dhtmlXGridCell.js *). There are several more files in the dhtml*.js category: dhtmlxgrid.js, dhtmlXGrid_skins.css, dhtmlXGrid_dhx_web.cs, dhtmlXGrid.css
- A strict check for semiduplicates for other filenames is required. A quick revealed >25 warnings from unzip.

**Comment by edsz14** [comment]
Be careful with this. Windows considers mixed case filenames without discerning upper or lower-case. However the PHP-scripts should preserve the mixed-case for filenames in order to be usable in both Windows and Linux. I remember specifically having worked out all references to files in the dhtmlXgrid libraries, a lot of work, in order to make them work in both environments. So it should always work in Windows but in Linux only if the mixed-case spellings are correct with the filenames.

**Comment by fho4abcd** [comment]
not everything resolved. More actions necessary

**Comment by edsz14** [comment]
You mean 'not everything resolved' re filenames with mixed case ? Then there should be things working in Windows but not in Linux. Examples of such cases ?

**Comment by fho4abcd** [comment]
Direct answer: Cloning ABCD2 with Github Desktop determines currently changed file
...www\htdocs\central\dataentry\js\dhtml_grid\dhtmlXGridCell.js
The upper and lowercase versions are very different. This implies that IF this file is not used we have a correct working application. IF the download sequence gives by accident the correct version we have also a correct working application.
Download and unpack a zip from gives also
- several "'bases-examples_Linux/secs-web/cityLib/titlePlus.* <=> titleplus.*
- htdocs\central\dataentry\js\dhtml_grid\dhtmlXGridCell.js <=>dhtmlxgridCell.js (different, used in circulation)
- www\htdocs\central\dataentry\js\dhtml_grid\dhtmlXGrid.js <=> dhtmlxgrid.js (equal, one should be deleted)
- 
The application might work but which version will be updated in Git if changed by Github Desktop?

**Comment by edsz14** [comment]
So you mean that during the cloning itself filenames are changed re mixed case, but even in an inconsistent way ? E.g. how can 
dhtmlXGridCell.js become dhtmlxgridCell.js ? Note that not all uppercase is changed to lowercase, something which would still be manageable because consistent if it would have changed to dhtmlxgridcell.js. 
I checked things and can see that some files are present in two varieties : central/dataentry/js/dhtmlxgrid.js exists as well as dhtmlXGrid.js. This doesn't harm (unless they are different inside the scripts) but of course we should remove the one without uppercase in the file-name. I suppose the one without uppercase does not observe case-sensitiveness and therefore cannot be used in Linux. So duplicates exist currently in the subfolder dataentry/js/dhtml_xgrid, but they have the same size and only differ in their use of mixed-case. 
We use these scripts from external JS libraries which are/were not maintained by ourselves. That has created already too many problems in ABCD.
I did a search inside the scripts for any call to lower-case only 'dhtmlxgrid' and found such in e.g. dhtmlXGrid.js at line 2785 :
document.body._dhtmlxgrid_onkeydown=true; 
where it refers to a property, not a filename.
No other calls are made to lower-case-only instances, except in the following : 

In dbadmin/fst.php however there is this (confusing because partly commented out) part to be corrected from line 44 : 
 
 <script language="JavaScript" type="text/javascript" src="../dataentry/js/dhtml_grid/dhtmlx.js"></script>
I don't see any calls to dhtmlXGrid elements in Circulation, by the way.
So, in my view we should (or could as it is not a real issue) delete the non-mixed-case scripts like dhtmlxgrid.js, but indeed should correct only one script as above : fst.php.     On Monday, February 8, 2021, 07:53:02 PM GMT+1, Fred <notifications@github.com> wrote:  
 
 


Direct answer: Cloning ABCD2 with Github Desktop determines currently changed file
...www\htdocs\central\dataentry\js\dhtml_grid\dhtmlXGridCell.js
The upper and lowercase versions are very different. This implies that IF this file is not used we have a correct working application. IF the download sequence gives by accident the correct version we have also a correct working application.
Download and unpack a zip from gives also
   
   - several "'bases-examples_Linux/secs-web/cityLib/titlePlus.* <=> titleplus.*
   - htdocs\central\dataentry\js\dhtml_grid\dhtmlXGridCell.js <=>dhtmlxgridCell.js (different, used in circulation)
   - www\htdocs\central\dataentry\js\dhtml_grid\dhtmlXGrid.js <=> dhtmlxgrid.js (equal, one should be deleted)
   - 

The application might work but which version will be updated in Git if changed by Github Desktop?

**Comment by fho4train** [comment]
Cloning with Github desktop and downloading and extracting a zip on
    Windows suffer both from the fact that filenames in Windows are not
    case-sensitive.
    A filename in Windows can have a mix of upper and lowercase
    characters and this is shown in the explorer.
    All copy actions in windows check for the filename of the target. A
    name with different case is in windows an equal file.
    
    Extracting a zip asks for overwrite or creating a copy (filename
    with a (1) )
    Cloning with Github desktop overwrites the original and, if the
    content is different, indicates that the file is changed.
    
    As far as I know it is unknown which file is first. Effects:
    - The windows copy will have less files as the GitHub content
    - The name of the file is the first downloaded file and the content
    is the second file
    - Uploading this file will overwrite the content of the original
    will with the new content
    Sufficient arguments for me to ban duplicate filenames, even if the
    content is equal.
    Filenames can retain a mixed upper/lowercase scheme, as long as
    there is no duplicate.
    I do not care if we retain the mixed case (probably the original) or
    a lowercase variant.
    
    Note that "central/dbadmin/*" uses now
    only dhtmlx.js.
        (=conform beta release for linux. BTW: dhtmlX.js==dhtmlx.js).(I
        have already deleted the comment parts in several files in order
        to minimize confusion)
    Note also that "central/circulation/typeofloans.php"
    specifies (not sure if really used):
    <script ....
      src="../js/dhtml_grid/dhtmlXCommon.js"></script>
      <script .... src="../js/dhtml_grid/dhtmlXGrid.js"></script>
      <script .... src="../js/dhtml_grid/dhtmlXGridCell.js"></script>
      <script .... src="../js/dhtml_grid/dhtmlXGrid_drag.js"></script>
      <script .... src="../js/dhtml_grid/dhtmlXGrid_excell_link.js"></script>
      <script .... src="../js/dhtml_grid/dhtmlXGrid_start.js"></script>
    This code cannot work because the path to the scripts is wrong and
    some scripts are not present at all in the beta release for linux.
    
    Switching to the original mixed filename is OK for me.
    
    
    
    
    Op 2021-02-09 om 10:11 schreef Egbert
      de Smet:
    
    
      
      So you mean that during the cloning itself filenames are changed
      re mixed case, but even in an inconsistent way ? E.g. how can 
      dhtmlXGridCell.js become dhtmlxgridCell.js ? Note that not all
      uppercase is changed to lowercase, something which would still be
      manageable because consistent if it would have changed to
      dhtmlxgridcell.js. 
      I checked things and can see that some files are present in two
      varieties : central/dataentry/js/dhtmlxgrid.js exists as well as
      dhtmlXGrid.js. This doesn't harm (unless they are different inside
      the scripts) but of course we should remove the one without
      uppercase in the file-name. I suppose the one without uppercase
      does not observe case-sensitiveness and therefore cannot be used
      in Linux. So duplicates exist currently in the subfolder
      dataentry/js/dhtml_xgrid, but they have the same size and only
      differ in their use of mixed-case. 
      We use these scripts from external JS libraries which are/were not
      maintained by ourselves. That has created already too many
      problems in ABCD.
      I did a search inside the scripts for any call to lower-case only
      'dhtmlxgrid' and found such in e.g. dhtmlXGrid.js at line 2785 :
      document.body._dhtmlxgrid_onkeydown=true; 
      where it refers to a property, not a filename.
      No other calls are made to lower-case-only instances, except in
      the following : 
      
      In dbadmin/fst.php however there is this (confusing because partly
      commented out) part to be corrected from line 44 : 
      
      <script language="JavaScript" type="text/javascript"
      src="../dataentry/js/dhtml_grid/dhtmlx.js"></script>
      I don't see any calls to dhtmlXGrid elements in Circulation, by
      the way.
      So, in my view we should (or could as it is not a real issue)
      delete the non-mixed-case scripts like dhtmlxgrid.js, but indeed
      should correct only one script as above : fst.php. On Monday,
      February 8, 2021, 07:53:02 PM GMT+1, Fred <notifications@github.com>
      wrote: 
      
      
      
      
      Direct answer: Cloning ABCD2 with Github Desktop determines
      currently changed file
      ...www\htdocs\central\dataentry\js\dhtml_grid\dhtmlXGridCell.js
      The upper and lowercase versions are very different. This implies
      that IF this file is not used we have a correct working
      application. IF the download sequence gives by accident the
      correct version we have also a correct working application.
      Download and unpack a zip from gives also
      
      - several "'bases-examples_Linux/secs-web/cityLib/titlePlus.*
      <=> titleplus.*
      - htdocs\central\dataentry\js\dhtml_grid\dhtmlXGridCell.js
      <=>dhtmlxgridCell.js (different, used in circulation)
      - www\htdocs\central\dataentry\js\dhtml_grid\dhtmlXGrid.js
      <=> dhtmlxgrid.js (equal, one should be deleted)
      - 
      
      The application might work but which version will be updated in
      Git if changed by Github Desktop?
      
      —
      You are receiving this because you commented.
      
      Reply to this email directly, view it on GitHub, or unsubscribe.
      —
        You are receiving this because you are subscribed to this
        thread.
        Reply to this email directly, view it on GitHub, or unsubscribe.
      [

{

"@context": "http://schema.org",

"@type": "EmailMessage",

"potentialAction": {

"@type": "ViewAction",

"target": "https://github.com/ABCD-DEVCOM/ABCD2/issues/7#issuecomment-775785408",

"url": "https://github.com/ABCD-DEVCOM/ABCD2/issues/7#issuecomment-775785408",

"name": "View Issue"

},

"description": "View this Issue on GitHub",

"publisher": {

"@type": "Organization",

"name": "GitHub",

"url": "https://github.com"

}

}

]

**Comment by edsz14** [comment]
The only real solution is to remove all case-insensitive (=lowercase only) duplicates of the scripts of dxhtmXGrid. As pointed out these are no longer called anywhere except in fst.php which can be easily adjusted for the filenames called from line 44.     On Tuesday, February 9, 2021, 12:32:57 PM GMT+1, Fred <notifications@github.com> wrote:  
 
 


 
 
 
 
 Cloning with Github desktop and downloading and extracting a zip on
 Windows suffer both from the fact that filenames in Windows are not
 case-sensitive.
 A filename in Windows can have a mix of upper and lowercase
 characters and this is shown in the explorer.
 All copy actions in windows check for the filename of the target. A
 name with different case is in windows an equal file.
 
 Extracting a zip asks for overwrite or creating a copy (filename
 with a (1) )
 Cloning with Github desktop overwrites the original and, if the
 content is different, indicates that the file is changed.
 
 As far as I know it is unknown which file is first. Effects:
 - The windows copy will have less files as the GitHub content
 - The name of the file is the first downloaded file and the content
 is the second file
 - Uploading this file will overwrite the content of the original
 will with the new content
 Sufficient arguments for me to ban duplicate filenames, even if the
 content is equal.
 Filenames can retain a mixed upper/lowercase scheme, as long as
 there is no duplicate.
 I do not care if we retain the mixed case (probably the original) or
 a lowercase variant.
 
 Note that "central/dbadmin/*" uses now
 only dhtmlx.js.
 (=conform beta release for linux. BTW: dhtmlX.js==dhtmlx.js).(I
 have already deleted the comment parts in several files in order
 to minimize confusion)
 Note also that "central/circulation/typeofloans.php"
 specifies (not sure if really used):
 <script ....
 src="../js/dhtml_grid/dhtmlXCommon.js"></script>
 <script .... src="../js/dhtml_grid/dhtmlXGrid.js"></script>
 <script .... src="../js/dhtml_grid/dhtmlXGridCell.js"></script>
 <script .... src="../js/dhtml_grid/dhtmlXGrid_drag.js"></script>
 <script .... src="../js/dhtml_grid/dhtmlXGrid_excell_link.js"></script>
 <script .... src="../js/dhtml_grid/dhtmlXGrid_start.js"></script>
 This code cannot work because the path to the scripts is wrong and
 some scripts are not present at all in the beta release for linux.
 
 Switching to the original mixed filename is OK for me.
 
 
 
 
 Op 2021-02-09 om 10:11 schreef Egbert
 de Smet:
 
 
 
 So you mean that during the cloning itself filenames are changed
 re mixed case, but even in an inconsistent way ? E.g. how can 
 dhtmlXGridCell.js become dhtmlxgridCell.js ? Note that not all
 uppercase is changed to lowercase, something which would still be
 manageable because consistent if it would have changed to
 dhtmlxgridcell.js. 
 I checked things and can see that some files are present in two
 varieties : central/dataentry/js/dhtmlxgrid.js exists as well as
 dhtmlXGrid.js. This doesn't harm (unless they are different inside
 the scripts) but of course we should remove the one without
 uppercase in the file-name. I suppose the one without uppercase
 does not observe case-sensitiveness and therefore cannot be used
 in Linux. So duplicates exist currently in the subfolder
 dataentry/js/dhtml_xgrid, but they have the same size and only
 differ in their use of mixed-case. 
 We use these scripts from external JS libraries which are/were not
 maintained by ourselves. That has created already too many
 problems in ABCD.
 I did a search inside the scripts for any call to lower-case only
 'dhtmlxgrid' and found such in e.g. dhtmlXGrid.js at line 2785 :
 document.body._dhtmlxgrid_onkeydown=true; 
 where it refers to a property, not a filename.
 No other calls are made to lower-case-only instances, except in
 the following : 
 
 In dbadmin/fst.php however there is this (confusing because partly
 commented out) part to be corrected from line 44 : 
 
 <script language="JavaScript" type="text/javascript"
 src="../dataentry/js/dhtml_grid/dhtmlx.js"></script>
 I don't see any calls to dhtmlXGrid elements in Circulation, by
 the way.
 So, in my view we should (or could as it is not a real issue)
 delete the non-mixed-case scripts like dhtmlxgrid.js, but indeed
 should correct only one script as above : fst.php. On Monday,
 February 8, 2021, 07:53:02 PM GMT+1, Fred <notifications@github.com>
 wrote: 
 
 
 
 
 Direct answer: Cloning ABCD2 with Github Desktop determines
 currently changed file
 ...www\htdocs\central\dataentry\js\dhtml_grid\dhtmlXGridCell.js
 The upper and lowercase versions are very different. This implies
 that IF this file is not used we have a correct working
 application. IF the download sequence gives by accident the
 correct version we have also a correct working application.
 Download and unpack a zip from gives also
 
 - several "'bases-examples_Linux/secs-web/cityLib/titlePlus.*
 <=> titleplus.*
 - htdocs\central\dataentry\js\dhtml_grid\dhtmlXGridCell.js
 <=>dhtmlxgridCell.js (different, used in circulation)
 - www\htdocs\central\dataentry\js\dhtml_grid\dhtmlXGrid.js
 <=> dhtmlxgrid.js (equal, one should be deleted)
 - 
 
 The application might work but which version will be updated in
 Git if changed by Github Desktop?
 
 —
 You are receiving this because you commented.
 
 Reply to this email directly, view it on GitHub, or unsubscribe.
 —
 You are receiving this because you are subscribed to this
 thread.
 Reply to this email directly, view it on GitHub, or unsubscribe.
 [

{

"@context": "http://schema.org",

"@type": "EmailMessage",

"potentialAction": {

"@type": "ViewAction",

"target": "https://github.com/ABCD-DEVCOM/ABCD2/issues/7#issuecomment-775785408",

"url": "https://github.com/ABCD-DEVCOM/ABCD2/issues/7#issuecomment-775785408",

"name": "View Issue"

},

"description": "View this Issue on GitHub",

"publisher": {

"@type": "Organization",

"name": "GitHub",

"url": "https://github.com"

}

}

]
